### PR TITLE
Count and log spectra that receive no prediction

### LINKED ISSUE #660: Improving logging wrt beam search failures
The Casanovo log file reports when spectra get skipped because they have too few peaks or out-of-range charge state.  However, some spectra fail to receive a prediction because the beam search terminates without a valid peptide because the predicted sequence is shorter than the minimum peptide length, an invalid N-terminal modification was predicted, or a padding token was predicted.  We should keep a count of these and report that count in the log file.

Attached is a sample MGF file that contains 10 spectra but for which Casanovo only makes one prediction.

[missing-predictions.zip](https://github.com/user-attachments/files/29700376/missing-predictions.zip)

### PR DESCRIPTION
**Pull request by ssiddhantsharma**
Closes #660.

Some spectra receive no prediction because beam search terminates without a valid peptide (e.g. the sequence is shorter than the minimum length, an invalid N-terminal modification, or a padding token). These were silently dropped, so the number of reported PSMs could be less than the number of input spectra with no indication why.

The number of such spectra is now counted in `predict_step` (when a spectrum yields no beam), reset at the start of prediction, and logged at the end of a run, so every spectrum is accounted for as either scored or skipped.

Added a test that mocks a mix of predicted and empty beam results and asserts the count and its reset.

Made with the help of AI tools; I have reviewed and take responsibility for all changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Spectra without valid peptide predictions are now skipped instead of producing invalid results.
  * Sequencing and database-search reports now count and display spectra lacking valid predictions.
  * Prediction statistics reset correctly at the start of each run and aggregate accurately across processing devices.
  * Runs with no missing predictions no longer produce unnecessary warnings, while missing-prediction counts are clearly logged.

### CODE REVIEW

File: casanovo/denovo/model.py [hunk 1]
```diff
@@ -1027,6 +1032,24 @@ def on_validation_epoch_end(self) -> None:
...
+        self.n_missing_predictions = 0
+
+    def on_predict_epoch_end(self) -> None:
+        """Log the number of spectra that did not receive a prediction."""
+        n_missing = int(
+            self.all_gather(
+                torch.tensor(self.n_missing_predictions, device=self.device)
+            ).sum()
+        )
+        if n_missing > 0 and self.trainer.is_global_zero:
```
Comment (bittremieux) [inline casanovo/denovo/model.py]: **todo:** This information should be passed to the [reporting function](https://github.com/Noble-Lab/casanovo/blob/main/casanovo/utils.py#L177) to synchronize end-of-run reporting in a consistent fashion.

Bonus points if we can harmonize logging of spectra skipped during preprocessing there as well.

**Review by bittremieux — CHANGES_REQUESTED** [review]

File: casanovo/denovo/model.py (hunk 1 above)
Comment (ssiddhantsharma) [reply casanovo/denovo/model.py]: Done in 6ac34b6. The missing-prediction count is now aggregated across devices in `on_predict_epoch_end` and passed to `log_annotate_report` it's reported as part of the end-of-run report now.

On the bonus, that count currently only surfaces as a warning from depthcharge's parser, which Casanovo doesn't capture as a value. I'd be glad to follow up on surfacing it through the same report in a separate change.

Comment (bittremieux) [reply casanovo/denovo/model.py]: Sounds good. It seems that there are a few new CodeRabbit comments. Can you resolve those and then tag me for a re-review afterwards? Thanks.

Comment (ssiddhantsharma) [reply casanovo/denovo/model.py]: Thanks! Addressed the CodeRabbit comments in 041d96b @bittremieux.
made `n_missing_predictions` keyword-only so positional `score_bins` callers keep working & suppressed the redundant empty-report warning

File: casanovo/utils.py [hunk 2]
```diff
@@ -206,9 +211,10 @@ def log_annotate_report(
     )
 
     if run_report is None:
-        logger.warning(
-            "No predictions were logged, this may be due to an error"
-        )
+        if n_missing_predictions == 0:
+            logger.warning(
+                "No predictions were logged, this may be due to an error"
```
Comment (bittremieux) [inline casanovo/utils.py]: **todo:** This warning message is incorrect. If `n_missing_predictions == 0`, that means that all spectra received a prediction. It can just be removed.

**Review by bittremieux — CHANGES_REQUESTED** [review]

File: casanovo/utils.py (hunk 2 above)
Comment (ssiddhantsharma) [reply casanovo/utils.py]: Removed the warning in 9521103 you're right, guarding it on `n_missing_predictions == 0` was contradictory.

**Comment by ssiddhantsharma** [comment]
@bittremieux ready for re-review. Removed the incorrect `No predictions were logged` warning in 9521103 (updated its unit test). The missing-prediction count is reported via `log_annotate_report` (end-of-run report) as you requested, and the earlier CodeRabbit comments were resolved in 041d96b.

File: casanovo/utils.py [hunk 3]
```diff
@@ -205,11 +210,7 @@ def log_annotate_report(
         score_bins=score_bins,
     )
 
-    if run_report is None:
```
Comment (bittremieux) [inline casanovo/utils.py:208]: **todo:** I'd still keep this original message in there, because explicitly warning the user that there are no predictions is useful. It was just that the previous check based on `n_missing_predictions` was incorrect.

**Review by bittremieux — CHANGES_REQUESTED** [review]

File: casanovo/utils.py (hunk 3 above)
Comment (ssiddhantsharma) [reply casanovo/utils.py:208]: Following your note, I kept the No predictions were logged warning and removed only the incorrect n_missing_predictions == 0 guard (https://github.com/Noble-Lab/casanovo/commit/0b60e6069ac07e2bd76c2c55d9d35cec31d45c76); the test now asserts the warning fires on an empty report. The missing-prediction count is reported via log_annotate_report, and the earlier CodeRabbit comments were resolved in https://github.com/Noble-Lab/casanovo/commit/041d96b695a98e23afce8dc1c6406be8b73e1baf.

**Review by bittremieux — APPROVED** [review]
Thanks!
